File: process_assets.py
import cv2
import numpy as np
import os

def process_sprite_sheet(input_path, output_path):
    print(f"Processing {input_path}...")
    
    # Read with alpha channel if present, otherwise BGR
    img = cv2.imread(input_path, cv2.IMREAD_UNCHANGED)
    if img is None:
        print(f"Failed to load {input_path}")
        return None
        
    # Convert to BGRA if it doesn't have an alpha channel
    if img.shape[2] == 3:
        img = cv2.cvtColor(img, cv2.COLOR_BGR2BGRA)
        
    # Find white pixels (background)
    # White is usually close to 255,255,255. We'll use a threshold.
    b, g, r, a = cv2.split(img)
    white_mask = (r > 240) & (g > 240) & (b > 240)
    
    # Set white background to transparent
    a[white_mask] = 0
    img = cv2.merge((b, g, r, a))
    
    # To find bounding boxes, we use the alpha channel
    # Find contours on the non-transparent pixels
    _, thresh = cv2.threshold(a, 10, 255, cv2.THRESH_BINARY)
    
    # Morphological close to connect slightly disconnected pixels (like wands)
    kernel = np.ones((5, 5), np.uint8)
    thresh = cv2.morphologyEx(thresh, cv2.MORPH_CLOSE, kernel)
    
    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    
    boxes = []
    for cnt in contours:
        x, y, w, h = cv2.boundingRect(cnt)
        # Filter out noise or text labels
        # Text is usually very wide but short, or just very small
        if w > 15 and h > 20 and w < 200 and h < 200:
            if float(w)/h < 3.0: # Not a long line of text
                boxes.append((x, y, w, h))
                
    if not boxes:
        print(f"No valid sprites found in {input_path}")
        return None
        
    # Sort boxes into rows based on Y center
    # We assume rows are separated by at least 30 pixels
    boxes.sort(key=lambda b: b[1] + b[3]/2) 
    
    rows = []
    current_row = []
    last_y_center = boxes[0][1] + boxes[0][3]/2
    
    for box in boxes:
        y_center = box[1] + box[3]/2
        if abs(y_center - last_y_center) > 30:
            rows.append(current_row)
            current_row = []
        current_row.append(box)
        last_y_center = y_center
        
    rows.append(current_row)
    
    # Sort each row by X coordinate
    for row in rows:
        row.sort(key=lambda b: b[0])
        
    # Find max grid dimensions
    max_w = max(b[2] for r in rows for b in r)
    max_h = max(b[3] for r in rows for b in r)
    num_rows = len(rows)
    num_cols = max(len(r) for r in rows)
    
    print(f"Detected {num_rows} rows, max {num_cols} cols.")
    print(f"Max sprite size: {max_w}x{max_h}")
    
    # Add padding to frame
    padding = 10
    frame_w = max_w + padding*2
    frame_h = max_h + padding*2
    
    # Create new blank transparent image
    out_img = np.zeros((frame_h * num_rows, frame_w * num_cols, 4), dtype=np.uint8)
    
    # Paste sprites into the grid centered
    for r_idx, row in enumerate(rows):
        for c_idx, box in enumerate(row):
            x, y, w, h = box
            
            # Extract sprite from original image (using actual pixels, not morphed)
            sprite = img[y:y+h, x:x+w]
            
            # Calculate paste position (center in grid cell)
            paste_x = c_idx * frame_w + (frame_w - w) // 2
            paste_y = r_idx * frame_h + (frame_h - h) // 2
            
            # Create a mask of the sprite's alpha channel
            sprite_alpha = sprite[:, :, 3] / 255.0
            
            # Alpha blend onto output (which is currently all 0s anyway, so we can just set it)
            out_img[paste_y:paste_y+h, paste_x:paste_x+w] = sprite
            
    cv2.imwrite(output_path, out_img)
    print(f"Saved cleaned sprite sheet to {output_path}")
    print(f"PHASER CONFIG: frameWidth: {frame_w}, frameHeight: {frame_h}")
    return frame_w, frame_h

def process_background(input_path, output_path):
    print(f"Processing background {input_path}...")
    img = cv2.imread(input_path)
    if img is None:
        print(f"Failed to load {input_path}")
        return
        
    # Resize to exactly 800x600 for the game
    img_resized = cv2.resize(img, (800, 600))
    cv2.imwrite(output_path, img_resized)
    print(f"Saved resized background to {output_path}")

if __name__ == "__main__":
    public_dir = r"C:\Users\HP\Downloads\RupertsDuelClub\public"
    
    blue_in = os.path.join(public_dir, "blue_wiz_sprite.png")
    blue_out = os.path.join(public_dir, "blue_wiz_clean.png")
    
    red_in = os.path.join(public_dir, "red_wiz_sprite.png")
    red_out = os.path.join(public_dir, "red_wiz_clean.png")
    
    bg_in = os.path.join(public_dir, "duel_bg.png")
    bg_out = os.path.join(public_dir, "duel_bg_clean.png")
    
    if os.path.exists(blue_in):
        process_sprite_sheet(blue_in, blue_out)
    else:
        print("Missing blue wizard")
        
    if os.path.exists(red_in):
        process_sprite_sheet(red_in, red_out)
    else:
        print("Missing red wizard")
        
    if os.path.exists(bg_in):
        process_background(bg_in, bg_out)
    else:
        print("Missing background")

File: process_assets.js
import { Jimp } from 'jimp';
import fs from 'fs';
import path from 'path';

const publicDir = 'public';

async function processSpriteSheet(filename, outname) {
    const filePath = path.join(publicDir, filename);
    if (!fs.existsSync(filePath)) {
        console.log("Missing " + filename);
        return;
    }
    console.log("Processing " + filename + "...");
    const img = await Jimp.read(filePath);
    
    const w = img.bitmap.width;
    const h = img.bitmap.height;
    
    // 1. Make white background transparent
    for (let y = 0; y < h; y++) {
        for (let x = 0; x < w; x++) {
            const hex = img.getPixelColor(x, y);
            const rgba = { r: (hex >>> 24) & 255, g: (hex >>> 16) & 255, b: (hex >>> 8) & 255, a: hex & 255 };
            if (rgba.r > 240 && rgba.g > 240 && rgba.b > 240) {
                img.setPixelColor(0x00000000, x, y);
            }
        }
    }
    
    // 2. Find bounding boxes using BFS on non-transparent pixels
    const visited = new Uint8Array(w * h);
    const boxes = [];
    
    for (let y = 0; y < h; y++) {
        for (let x = 0; x < w; x++) {
            const idx = y * w + x;
            if (visited[idx]) continue;
            visited[idx] = 1;
            
            const hex = img.getPixelColor(x, y);
            const rgba = { r: (hex >>> 24) & 255, g: (hex >>> 16) & 255, b: (hex >>> 8) & 255, a: hex & 255 };
            
            if (rgba.a > 10) {
                let minX = x, maxX = x, minY = y, maxY = y;
                const q = [[x, y]];
                let head = 0;
                
                while(head < q.length) {
                    const [cx, cy] = q[head++];
                    
                    const neighbors = [
                        [cx+1, cy], [cx-1, cy], [cx, cy+1], [cx, cy-1],
                        [cx+1, cy+1], [cx-1, cy-1], [cx+1, cy-1], [cx-1, cy+1]
                    ];
                    
                    for (const [nx, ny] of neighbors) {
                        if (nx >= 0 && nx < w && ny >= 0 && ny < h) {
                            const nidx = ny * w + nx;
                            if (!visited[nidx]) {
                                visited[nidx] = 1;
                                const nhex = img.getPixelColor(nx, ny);
                                const nrgba = { r: (nhex >>> 24) & 255, g: (nhex >>> 16) & 255, b: (nhex >>> 8) & 255, a: nhex & 255 };
                                if (nrgba.a > 10) {
                                    if (nx < minX) minX = nx;
                                    if (nx > maxX) maxX = nx;
                                    if (ny < minY) minY = ny;
                                    if (ny > maxY) maxY = ny;
                                    q.push([nx, ny]);
                                }
                            }
                        }
                    }
                }
                
                const bw = maxX - minX + 1;
                const bh = maxY - minY + 1;
                
                if (bw > 100 && bh > 100 && bw < 600 && bh < 600) {
                    if (bw / bh < 3.0) {
                        boxes.push({x: minX, y: minY, w: bw, h: bh});
                    }
                }
            }
        }
    }
    
    if (boxes.length === 0) {
        console.log("No valid sprites found in " + filename);
        return;
    }
    
    boxes.sort((a, b) => (a.y + a.h/2) - (b.y + b.h/2));
    
    const rows = [];
    let currentRow = [];
    let lastY = boxes[0].y + boxes[0].h/2;
    
    for (const box of boxes) {
        const y = box.y + box.h/2;
        if (Math.abs(y - lastY) > 60) {
            rows.push(currentRow);
            currentRow = [];
        }
        currentRow.push(box);
        lastY = y;
    }
    rows.push(currentRow);
    
    for (const row of rows) {
        row.sort((a, b) => a.x - b.x);
    }
    
    let maxW = 0, maxH = 0;
    let maxCols = 0;
    for (const row of rows) {
        if (row.length > maxCols) maxCols = row.length;
        for (const box of row) {
            if (box.w > maxW) maxW = box.w;
            if (box.h > maxH) maxH = box.h;
        }
    }
    
    const padding = 10;
    const frameW = maxW + padding * 2;
    const frameH = maxH + padding * 2;
    const numRows = rows.length;
    
    console.log(`${filename}: Detected ${numRows} rows, max cols ${maxCols}`);
    console.log(`${filename}: Max size: ${maxW}x${maxH}, Frame Size: ${frameW}x${frameH}`);
    
    const outImg = new Jimp({ width: maxCols * frameW, height: numRows * frameH, color: 0x00000000 });
    
    for (let r = 0; r < rows.length; r++) {
        for (let c = 0; c < rows[r].length; c++) {
            const box = rows[r][c];
            const sprite = img.clone();
            sprite.crop({ x: box.x, y: box.y, w: box.w, h: box.h });
            
            const pasteX = c * frameW + Math.floor((frameW - box.w) / 2);
            const pasteY = r * frameH + Math.floor((frameH - box.h) / 2);
            
            outImg.composite(sprite, pasteX, pasteY);
        }
    }
    
    await outImg.write(path.join(publicDir, outname));
    console.log(`Saved ${outname}. PHASER CONFIG: ${frameW}x${frameH}`);
    fs.writeFileSync(path.join(publicDir, outname + '.json'), JSON.stringify({frameWidth: frameW, frameHeight: frameH, rows, maxCols}));
}

async function processBg() {
    const filePath = path.join(publicDir, 'duel_bg.png');
    if (!fs.existsSync(filePath)) return;
    console.log("Processing bg...");
    const img = await Jimp.read(filePath);
    img.resize({ w: 800, h: 600 });
    await img.write(path.join(publicDir, 'duel_bg_clean.png'));
    console.log("Saved bg.");
}

async function main() {
    await processSpriteSheet('blue_wiz_sprite.png', 'blue_wiz_clean.png');
    await processSpriteSheet('red_wiz_sprite.png', 'red_wiz_clean.png');
    await processBg();
}

main().catch(console.error);
